<img src="https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png" style="float: left; margin: 15px;">

# Intro to Data Cleaning

Week 2 | Lesson 2.3

<img src="https://snag.gy/ywU34V.jpg" width="250">

### LEARNING OBJECTIVES
*After this lesson, you will be able to:*
- Inspect data types
- Clean up a column using df.apply()
- Know what situations to use .value_counts() in your code

### LESSON GUIDE
| TIMING  | TYPE  | TOPIC  |
|:-:|---|---|
| 5 min  | [Introduction](#introduction)   | Inpsect data types, df.apply(), .value_counts()  |
| 20 min  | [Demo /Guided Practice](#demo)  | Inpsect data types |
| 20 min  | [Demo /Guided Practice](#demo)  | df.apply() |
| 20 min  | [Demo /Guided Practice](#demo)  | .value_counts() |
| 20 min  | [Independent Practice](#ind-practice)  |   |
| 5 min  | [Conclusion](#conclusion)  |   |

<a name="introduction"></a>
## Introduction: Topic (5 mins)

Since we're starting to get pretty comfortable with using pandas to do EDA, let's add a
couple more tools to our toolbox. 

The main data types stored in pandas objects are **float, int, bool, datetime64, timedelta, 
category,** and **object****. 

df**.apply()** will apply a function along any axis of the DataFrame. We'll see it in action below. 

pandas.Series**.value_counts** returns _Series_ containing counts of unique values. The resulting 
Series will be in descending order so that the first element is the most frequently-occurring 
element. Excludes NA values.

[Pandas: dtypes](http://pandas.pydata.org/pandas-docs/stable/pandas.pdf)

[Pandas Series: value_counts](http://nullege.com/codes/search/pandas.Series.value_counts)


### Common Operations

- **float**: precision specific math operations
- **int**: mathatical operations using whole numbers
- **bool**: control flow conditions
- **datetime64**: resampling, slicing/selection, frequency back/front filling, on date range
- **timedelta**: date comparisons
- **category**: is like a set type, allowing to capture categorical day as a set, with ordinal
- **object**: all types can be represented as an object but math, date operations are not possible.  Limited control flow possibilities unless you are comparing strings.

<a name="Inpsect data types "></a>
## Demo /Guided Practice: Inspect data types  (20 mins)

Let's create a small dictionary with different data types in it. 

> [demo code](https://github.com/generalassembly-studio/dsi-course-materials/blob/W2-L2.3/curriculum/04-lessons/week-02/2.3-lesson/code/W2%20L2.3%20Intro%20to%20Data%20Cleaning%20demo%20code.ipynb)
can be found in the code folder and contains all the code in this lesson in a Jupyter
notebook. Follow along or create a new notebook.


### Import Pandas + Numpy

In [2]:
import pandas as pd
import numpy as np

### Create Test Data

In [3]:
test_data = dict( 
    A = np.random.rand(3),
    B = 1,
    C = 'foo',
    D = pd.Timestamp('20010102'),
    E = pd.Series([1.0]*3).astype('float32'),
    F = False,
    G = pd.Series([1]*3,dtype='int8')
)

In [4]:
test_data

{'A': array([ 0.72984486,  0.81514571,  0.54726853]),
 'B': 1,
 'C': 'foo',
 'D': Timestamp('2001-01-02 00:00:00'),
 'E': 0    1.0
 1    1.0
 2    1.0
 dtype: float32,
 'F': False,
 'G': 0    1
 1    1
 2    1
 dtype: int8}

### Create our DataFrame

In [5]:
dft = pd.DataFrame(test_data)
dft

A  B    C          D    E      F  G
0  0.729845  1  foo 2001-01-02  1.0  False  1
1  0.815146  1  foo 2001-01-02  1.0  False  1
2  0.547269  1  foo 2001-01-02  1.0  False  1

In [14]:
dft.dtypes

A           float64
B             int64
C            object
D    datetime64[ns]
E           float32
F              bool
G              int8
dtype: object

**What might we expect dtypes in the case of mixed type values in a single dimension?**

ie:  [2, 3, 4, 5, 6, 7, 8.9]

If a pandas object contains data multiple dtypes IN A SINGLE COLUMN, the dtype of the column will be chosen to accommodate all of the data types (object is the most general).

### Ints are cast to floats

In [18]:
pd.Series([1, 2, 3, 4, 5, 6.])

0    1.0
1    2.0
2    3.0
3    4.0
4    5.0
5    6.0
dtype: float64

### String elements are cast to ``object`` dtype

In [56]:
pd.Series([1, 2, 3, 'foo'])

0      1
1      2
2      3
3    foo
dtype: object

In [63]:
dft.get_dtype_counts().astype(list)

bool              1
datetime64[ns]    1
float32           1
float64           1
int64             1
int8              1
object            1
dtype: object

### With a partner, take 3 minutes to discus:

*Without* running this code with a Python interpreter, what types would you expect the common `dtype` to be selected?

    [1, 3, 9, .33, False, '03-20-1978', np.arange(22)]



You can do a lot more with dtypes.  Check out 
[Pandas Documentation](http://pandas.pydata.org/pandas-docs/stable/pandas.pdf).

## Why do you think it might be important to know what kind of dtypes you're working with? 

<a name=" df.apply()"></a>
## Demo /Guided Practice:  df.apply() (20 mins)

Generally, df.apply(), will apply a singlular function to every cell of the dataframe you use it with.  

Conversely: df.map(), is available when you only want to work with a single dimension of your dataset, ie:  df['a'].map(my_func)

In [28]:
# Create some more test data
df = pd.DataFrame(np.random.randn(5, 4), columns=['a', 'b', 'c', 'd'])
df

a         b         c         d
0 -1.369438  0.080447 -1.224570  0.422078
1 -0.765415  0.407583  0.385544  0.834827
2  1.758356  0.449190 -1.435722 -0.234399
3  1.172705 -1.596128 -0.821374  0.892523
4  0.026303 -0.749009 -0.940387  0.117611

### Some Examples

In [30]:
# square root ALL CELLS (NaN == Not a Number)
df.apply(np.sqrt)

a         b         c         d
0       NaN  0.283631       NaN  0.649675
1       NaN  0.638422  0.620922  0.913689
2  1.326030  0.670216       NaN       NaN
3  1.082915       NaN       NaN  0.944734
4  0.162181       NaN       NaN  0.342944

_Note: Illustrate with whiteboard DataFrame, with blank axis labels._

### Apply method to only one axis, 0 (columns)
ie:  `df['a'] == [-1.369438  , -0.76541512,  1.75835588,  1.17270527,  0.02630271]`

In [31]:
df.apply(np.mean, axis=0)

a    0.164502
b   -0.281584
c   -0.807302
d    0.406528
dtype: float64

### Apply method to only axis 1 (rows)
This is what the data slice would look like if we were to select only rows.  Here is the slice of the first row that would be affected with axis #1 with .apply():

`df.iloc[0].values == [-1.369438  ,  0.0804468 , -1.22457047,  0.42207757]`

_We are calculating the mean of lists of "rows", not "columns"._

In [32]:
df.apply(np.mean, axis=1)

0   -0.522871
1    0.215635
2    0.134356
3   -0.088068
4   -0.386371
dtype: float64

### Further Reading

For more advanced `.apply` usage, check out these links:

["Why Not"'s Gist Examples](https://gist.github.com/why-not/4582705)

[Chris Albon's Map + Apply Examples](http://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html)


### **Check:** How would find the std of the columns and rows? 

<a name=".value_counts()"></a>
## Demo /Guided Practice: .value_counts() (< 20 mins)

Why is this important?  Basically, this tells us the count of unique values that exist.  It's helpful to identify anything unexpected.  Looking at value_counts(), per series, can give us a quick overview of values expressed in our data.

 - Strings inside of mostly numeric / continious data
 - Non-numeric values
 - General counts of values that we might expect to see
 - Most common / least common values

Let's create some random data

In [10]:
data = np.random.randint(0, 7, size = 50)
data

array([2, 6, 0, 3, 0, 6, 0, 1, 1, 0, 6, 5, 4, 6, 6, 6, 2, 5, 3, 3, 5, 4, 6,
       6, 2, 6, 4, 1, 0, 3, 1, 6, 3, 0, 2, 3, 3, 4, 5, 5, 4, 2, 1, 2, 1, 1,
       3, 3, 5, 1])

In [8]:
s = pd.Series(data)
s.head()

0    5
1    3
2    5
3    4
4    5
dtype: int64

In [30]:
# The counts of each number that occurs in our array is listed
pd.value_counts(s)

5    14
2    10
4     9
3     7
0     4
6     3
1     3
dtype: int64

<a name="ind-practice"></a>
## Independent Practice: Topic (20+ minutes)
- Use the sales.csv data set, we've seen this a few times in previous lessons
- Inspect the data types
- You've found out that all your values in column 1 are off by 1. Use df.apply to add 1 to column 1 of the dataset
- Use .value_counts to count the values of 1 column of the dataset

**Bonus** 
- Add 3 to column 2
- Use .value_counts for each column of the dataset

**Bonus Bonus -- COMPLETELY OPTIONAL!!!**
<img src="http://vignette3.wikia.nocookie.net/erbparodies/images/a/a3/Troll_Based_On.png/revision/latest?cb=20151109194505" style="width: 100px;">

Ruining data should give you a better sense of how to clean it.  Don't feel like you need to attempt this as it's completely optional and it's meant to be _extranious_.  Real-life datasets will not be like this.  The solution isn't as important as the process and thinking behind your approach.  Another way you might want to try to do this, is map out how you would do this with pseudo code with a step-by-step plan, without actually coding anything.

- Add an extra column to your dataframe that is a copy of an existing column with continious data
    - Randomly change the value of continious data cells within it to the following:
      - NaN
      - A blank string
      - A numeric string
      - The same value
    - Report value_counts post-"random data troll" processing. Does it seem random?
    - Convert blank strings and NaN values to float(0)
    - Convert numeric strings to floats with 2f precision
    - Divide by 2 if cell value is prime, use remainder as value
    - Post solution as Gist with comments to Slack

In [29]:
import pandas as pd
import numpy as np
sales_path = '/Users/tlee010/desktop/dsi-sf-2-timdavidlee/datasets/sales_data_simple/sales.csv'
sales_data = pd.read_csv(sales_path)
sales_data.info()
print sales_data.head(5)
def addone(in_data):
    return in_data+1
def addtwo(in_data):
    return in_data+2
sales_data_adj = sales_data.copy()
sales_data_adj.iloc[:,0] = sales_data.iloc[:,0].map(addone)
sales_data_adj.iloc[:,1] = sales_data.iloc[:,1].map(addtwo)
print sales_data_adj.head(5)

print sales_data_adj['volume_sold'].value_counts()
print sales_data_adj['2015_margin'].value_counts()
print sales_data_adj['2015_q1_sales'].value_counts()
print sales_data_adj['2016_q1_sales'].value_counts()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
volume_sold      200 non-null float64
2015_margin      200 non-null float64
2015_q1_sales    200 non-null float64
2016_q1_sales    200 non-null float64
dtypes: float64(4)
memory usage: 6.3 KB
   volume_sold  2015_margin  2015_q1_sales  2016_q1_sales
0    18.420760    93.802281      337166.53      337804.05
1     4.776510    21.082425       22351.86       21736.63
2    16.602401    93.612494      277764.46      306942.27
3     4.296111    16.824704       16805.11        9307.75
4     8.156023    35.011457       54411.42       58939.90
   volume_sold  2015_margin  2015_q1_sales  2016_q1_sales
0    19.420760    95.802281      337166.53      337804.05
1     5.776510    23.082425       22351.86       21736.63
2    17.602401    95.612494      277764.46      306942.27
3     5.296111    18.824704       16805.11        9307.75
4     9.156023    37.011457       54411.42       58939.90
7.841363

<a name="conclusion"></a>
## Conclusion (5 mins)
So far we've used pandas to look at the head and tail of a data set. We've also taken a look at summary stats and different
types of data types. We've selected and sliced data too. Today we added inspecting data types, df.apply, .value_counts to
our pandas arsenal. Nice!